### DISCUSSION TITLE
Some questions

### DISCUSSION THREAD

**Discussion by TomislavTom** [Try/OpenGothic#394] [General]
Hello, I finally tried it out and I'm very impressed!

But I have some questions:
1. Are there any graphics options?
3. How do I display my fps?
4. Can I somehow turn off the light blue distant "fog"? I know it's vanilla-like but imho it does'nt make sense when you can render the whole map.
5. I installed Reportversion 2.6.0.0, Playerkit 2.6f and Systempack 1.8. Does this change anything? Are they even active?
6. How exactly do use the "-game:<modfile.init>" command? I tried the following to play with l'hiver:
a) "-game:<LHE204_DE.init>" 
b) "-game:<LHE204_DE.ini>" 
c) "-game:LHE204_DE.init" 
d) "-game:LHE204_DE.ini" 
but none worked.

Thanks in advance

**Comment by Try** [comment]
Hi, @TheDovahkiin97 !

> Are there any graphics options?

"cloud shadows" is remapped to SSAO
"radial fog" is remapped to light-shaft scattering
`-rt 0`, in command line is to disable ray-tracing
`-ms 0`, in command line is to disable mesh-shading (this is not a real graphics feature, more to addres possible issues)

> How do I display my fps?

Same as vanilla game : `marvin` -> `toggle frame`

> Can I somehow turn off the light blue distant "fog"?

Hm, now sure what this is about. There is some fog - sure, but shouldn't be a much. Can you share screenshoot, to illustrate the problem?

> I installed Reportversion 2.6.0.0, Playerkit 2.6f and Systempack 1.8. Does this change anything? Are they even active?

Those patches are design to be injected into original gorhic2.exe 32-bit binary and fix issue that are not in OpenGothic to begin with.
Having them around should cause no effect.

> How exactly do use the "-game:<modfile.init>" command?

Those are primarily for GothicStarter. So, if you replace original Gothic2.exe with OpenGothic (and rename exe), than starter can communicate with to opengothic and notify what mod is needed.

As for manual usage, here is command line that I've been using for KoM:
`-g "C:\Program Files (x86)\Steam\steamapps\common\TheChroniclesOfMyrtana" -game:TheChroniclesOfMyrtana.ini ` 
(sorry, have no L'hiver as mode, but should be same)

**Reply by TomislavTom** [reply]
Thank you!
The -game command worked and I got into the main menu but there I experienced a very unexpected problem:
1. I press start new game
2. it loads to approximately a quarter
3. It jumps back to the main menu where I can repeat :(

I also tried a version where I just have to drop a .vdf file into /Data but the game does'nt even start with this and Open Gothic...

I still wanted to try the following:
> Those are primarily for GothicStarter. So, if you replace original Gothic2.exe with OpenGothic (and rename exe), than starter can communicate with to opengothic and notify what mod is needed.

But when I replace the original Gothic2.exe with your Gothic2Notr.exe, rename it to Gothic2.exe and use the GothicStarter.exe to start the game or a mod I get multiple missing dll errors and nothing happens :(

>Can you share screenshoot, to illustrate the problem?

![image](https://user-images.githubusercontent.com/31317447/215286237-c4399d19-dc54-4ca7-bd73-ae0f07422cd7.png)
I will just use your Screenshot. I want the blue tint to disappear. I want distant objects to have the same colors as if they were right next to me.

Thanks in advance!

**Reply by TomislavTom** [reply]
> It jumps back to the main menu where I can repeat :(

I think I figured out the issue myself. The log.txt says it detected Ikarus/LeGo which are not yet supported by Open Gothic.
What do you think how long it will take you to make it compatible?

> I also tried a version where I just have to drop a .vdf file into /Data but the game does'nt even start with this and Open Gothic...

```
OpenGothic v1.0.1891

---crashlog(std::terminate)---
GPU: ?
0x00007ff738c1f456: dbg::call_stack<64u>::collect(unsigned int) in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738a9f8bc: CrashLog::dumpStack(char const*) in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738a9fe18: terminateHandler() in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ffd578663f6: ZN10__cxxabiv111__terminateEPFvvE in C:\Games\Gothic 2 Gold Edition\Open Gothic\libstdc++-6.dll
0x00007ffd57962d43: ZSt9terminatev in C:\Games\Gothic 2 Gold Edition\Open Gothic\libstdc++-6.dll
0x00007ffd5796bf56: _cxa_throw in C:\Games\Gothic 2 Gold Edition\Open Gothic\libstdc++-6.dll
0x00007ff738c88d56: phoenix::symbol* phoenix::script::_check_member<phoenix::c_fight_ai, phoenix::c_fight_ai_move, 6>(std::basic_string_view<char, std::char_traits<char> >, std::type_info const*) in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738b62415: phoenix::register_all_script_classes(phoenix::script&) in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738a10d85: Gothic::createPhoenixVm(std::basic_string_view<char, std::char_traits<char> >) in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff7389ca02e: FightAi::FightAi() in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738a082b4: Gothic::setupGlobalScripts() in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff738cf0d58: main in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff7389a13ae: __tmainCRTStartup in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ff7389a14c6: WinMainCRTStartup in C:\Games\Gothic 2 Gold Edition\Open Gothic\Gothic2Notr.exe
0x00007ffdc80b26bd: BaseThreadInitThunk in C:\Windows\System32\KERNEL32.DLL
0x00007ffdc94edfb8: RtlUserThreadStart in C:\Windows\SYSTEM32\ntdll.dll
```

**Reply by Try** [reply]
> The -game command worked and I got into the main menu but there I experienced a very unexpected problem

Can you share download link for Exact version of mod that you using? I have an old `0.9a` - works just fine

> I will just use your Screenshot. I want the blue tint to disappear.

Ah, basically you want to have no fog :D

>  multiple missing dll errors and nothing happens

Do you mean Tempest.dll/C++ runtime? Just copy them from archive as well

> What do you think how long it will take you to make it compatible?

We have a ticket #231, that is subdivided in 136 tasks - basically game needs to emulate x86 mmu in order to have partial ikarus support

**Reply by TomislavTom** [reply]
> Can you share download link for Exact version of mod that you using? I have an old 0.9a - works just fine

1.  The one not loading past the menu: LHiver204_DE_22-06-20.exe at https://www.worldofgothic.de/dl/download_556.htm
2. The one not starting at all https://drive.google.com/file/d/18RO757hZfcywev7DIBvR8FQr-W2L9C5d/view and https://forum.worldofplayers.de/forum/threads/1573073-RELEASE-Gothic-2-die-Nacht-des-Raben-Buddygoths-Lhiver-deutsche-Version
3. Perhaps ignore both of them because I discovered Returning 2.0 but it doesn't work either. It replaces the original Gothic2.exe but Open Gothic starts the old one: https://forum.worldofplayers.de/forum/threads/1555665-RELEASE-Returning-v2-0-Buddygoth-und-Tentarr%C2%B4s-Version

(I'm open to other mod suggestions for a nice 2023 Gothic experience playthrough)

> Ah, basically you want to have no fog :D

Yes please :D

> Do you mean Tempest.dll/C++ runtime? Just copy them from archive as well

It worked but I have a new error: 
invalid gothic path: ""

where am I supposed to edit this path?

> basically game needs to emulate x86 mmu

I have no idea what this means but it sounds like a lot of work :(

**Reply by Try** [reply]
> LHiver204_DE_22-06-20.exe

I've give it a run locally - seems mod uses more `move` slots in AI than allowed.
CC @lmichaelis, apparently in scripts there is:
```
CLASS C_FightAI 
{
	VAR INT move[7]; // instead of 6
};
```

>  Returning 2.0

Haven't look much into, but that mod seems to be worst-case of DMA, so yeah 'just' need to emulate x86 mmu :/

> invalid gothic path: ""

Hm, look like a bug - will investigate

**Reply by TomislavTom** [reply]
Okay no problem I'll wait ✌️

But will you implement an option to turn off fog?

**Reply by lmichaelis** [reply]
> > LHiver204_DE_22-06-20.exe
> 
> I've give it a run locally - seems mod uses more `move` slots in AI than allowed. CC @lmichaelis, apparently in scripts there is:
> 
> ```
> CLASS C_FightAI 
> {
> 	VAR INT move[7]; // instead of 6
> };
> ```


Hm, there is a solution but it's not very elegant. First though, I re-checked Gothic's and Gothic II's `FIGHTAI.DAT` and I can say with certainty that 6 is the correct number of `C_FIGHTAI.MOVE`s:

<details>
<summary><b>Gothic 1</b></summary>

```
00000012 <C_FIGHTAI>
	Name: C_FIGHTAI
	Address: ffffffff
	Size: 1
	Type: class
	Generated: No
	Flags:
		Const: No
		Return: No
		Member: No
		Extern: No
		Merged: No
	File Index: 0
	Line Start: 26
	Line Count: 4
	Char Start: 1069
	Char Count: 51
	Class Size: 24
	Definition:
		class C_FIGHTAI {
			var int MOVE[6];
		};
```
</details>

<details>
<summary><b>Gothic 2</b></summary>

```
00000014 <C_FIGHTAI>
	Name: C_FIGHTAI
	Address: ffffffff
	Size: 1
	Type: class
	Generated: No
	Flags:
		Const: No
		Return: No
		Member: No
		Extern: No
		Merged: No
	File Index: 0
	Line Start: 33
	Line Count: 4
	Char Start: 1474
	Char Count: 51
	Class Size: 24
	Definition:
		class C_FIGHTAI {
			var int MOVE[6];
		};
```
</details>

L'Hiver (at least in this version) actually does use 7 `FIGHTAI.MOVE`s. That's a rather weird choice if you ask me.

<details>
<summary><b>L'Hiver</b></summary>

```
00000014 <C_FIGHTAI>
	Name: C_FIGHTAI
	Address: ffffffff
	Size: 1
	Type: class
	Generated: No
	Flags:
		Const: No
		Return: No
		Member: No
		Extern: No
		Merged: No
	File Index: 0
	Line Start: 22
	Line Count: 4
	Char Start: 554
	Char Count: 40
	Class Size: 28
	Definition:
		class C_FIGHTAI {
			var int MOVE[7];
		};
```
</details>

A solution could be to actually just up the number of `MOVE`s supported by _phoenix_ since scripts that don't use that many just don't care. What do you think @Try? Should I just up that number in _phoenix_ or should I rather provide a way for you to alter it using `#define`s or templates?

**Reply by Try** [reply]
@lmichaelis , my assumption is that Vanilla can't really work with this case eider, and changing layout of engine-classes is UB.
For this case, trimming script array size and print warning is probably good solution.

**Reply by Try** [reply]
> But will you implement an option to turn off fog?

Well, eventually :)
Probably will have to introduce some config file like OpenGothic.ini, since we running out of traditional menu settings

**Reply by lmichaelis** [reply]
> my assumption is that Vanilla can't really work with this case eider

Yeah that does seem to be the case, poking at a decomp.

> trimming script array size and print warning is probably good solution.

That might work. phoenix will fail accessing the element at that index and throw an `illegal_index_access` exception which can be handled by the VM exception handler (which is current set to the default `lenient_vm_exception_handler`).

https://github.com/Try/OpenGothic/blob/1006d12ff6724657bbd1a76d3b9d357f99d689b0/game/game/gamescript.cpp#L68-L70

**Comment by toolatetotheparty** [comment]
> 3\. Can I somehow turn off the light blue distant "fog"? I know it's vanilla-like but imho it does'nt make sense when you can render the whole map.

The tint of the fog, ideally, should take the color of the sky, so during the daylight it hould have a blue tint. It would be nice to adjust the "fog-min" and "fog-max" distance values, though, since the scale of the world is small and aerial perspective should artistically scale with it.

**Reply by Try** [reply]
> should take the color of the sky,

Currently fog is based on UE 4.27 way of doing physical fog, so most stuff from their page is applicable:
https://docs.unrealengine.com/4.27/en-US/BuildingWorlds/FogEffects/VolumetricFog/

Currently, in OpenGothic
```
const float viewDistanceScale      = 20 - this is to take care of small world size

const vec3  rayleighScatteringBase = vec3(0.175, 0.409, 1.0) / 1e6;
const float rayleighAbsorptionBase = 0.0;

const float mieScatteringBase      = 3.996 / 1e6;
const float mieAbsorptionBase      = 4.40  / 1e6;

const vec3  ozoneAbsorptionBase    = vec3(0.650, 1.881, .085) / 1e6;
```

Scattering term are also multiplied by cloud concentration on a level to address Valey of mines look.
